TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+* Four dashboards (LabView dashboard, SmartDashboard, Shuffleboard, Glass)
+* Two logging solutions (Shuffleboard recordings, on-RIO logging)
+* A declarative object-oriented telemetry API (Sendable)
+* A code-bypassing "debug mode" for sensors and actuators (LiveWindow)
+
+While the WPILib telemetry ecosystem is reasonably functional (especially when combined with some amount of external tooling), it can be confusing from the outside due to the large amount of duplication in the above tools ("which of the four dashboards should I use and why?").  Additionally, as a result of having emerged from a shared set of "ancestral" features, many of the above tools have overlapping implementations that interact in surprising and sometimes undocumented ways.
+
+Healthy software development generally follows two alternating phases: expansion and consolidation.  The WPI telemetry ecosystem has grown through steady expansion into a substantial and powerful set of features.  It is now in need of a consolidation phase to reduce the total API surface area.  Consolidating and refactoring the existing implementations will serve two purposes: it will provide new users with a clearer and more-intuitive "happy path" for their robot telemetry, and it will reduce the maintenance burden on on the WPILib development team by disentangling the implementations of unrelated API features and reducing the total volume of maintained code.
+
+### Too many dashboards
+
+The current API presents new users with a number of pain points.  Firstly, users are presented with a choice of four different dashboards - each with slightly-different featuresets, but without enough distinguishing features to clearly guide users as to which dashboard is useful in which situation.
+
+In particular:
+
+* The LabView dashboard and SmartDashboard are largely legacy features that have been totally supplanted in functionality by Shuffleboard and Glass.  Neither is likely to be easily maintainable in the event of a major breakage.  Both options are substantially clunkier and less-functional than the alternatives, and teams which choose these dashboards are probably worse-off for the choice.  Is continued support for legacy team-side dashboard code really worth this cost, especially given the tendency of FRC teams to rewrite most of their code anually?
+* ShuffleBoard is clearly superior to Glass as an operating dashboard, and has many features (robot-code-driven tabs and layouts, data recording and playback) that make it appealing as the "default" robot dashboard.  However, it is already in danger of slipping into the "legacy code" state of the above two tools due to the migration of developer effort to Glass.  This could eventually cost teams if support for the dashboard becomes difficult to sustain while no other dashboard option offers a similar featureset.
+* Glass is a *fantastic* tool for real-time plotting and signal analysis, but currently lacks a robust featureset to be usable as a dashboard.
```

Comment:
X-Y plots would be a nice addition to time-series plots.

Reply:
Please open an issue on the allwpilib repo for Glass feature requests.